cleanlab/dataset.py

```diff
@@ -0,0 +1,416 @@
+# (at your option) any later version.
+#
+# cleanlab is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with cleanlab.  If not, see <https://www.gnu.org/licenses/>.
+
+
+"""Dataset Module
+Supports dataset-level and class-level automated quality, including finding which classes should
+be merged with other classes in your dataset, which classes should be removed,
+and which classes tend to be annotated most/least correctly overall.
+"""
+
+import numpy as np
+import pandas as pd
+from cleanlab.count import estimate_joint
+
+
+def rank_classes_by_label_quality(
+    labels=None,
+    pred_probs=None,
+    *,
+    class_names=None,
+    num_examples=None,
+    joint=None,
+    confident_joint=None,
+    multi_label=False,
+):
+    """Returns a Pandas DataFrame with all classes and three overall class label quality scores
+     (details about each score are listed in the Returns parameter). By default, classes are ordered
+     by "Overall Label Noise Score" descending.
+
+    Score values are unnormalized and may tend to be very small. What matters is their relative
+    ranking across the classes.
+
+    Parameters
+    ----------
+    For parameter info, see the docstring of `dataset.find_overlapping_classes`
+
+    Returns
+    -------
+    pd.DataFrame
+        A Pandas DataFrame with cols "Class Index", "Overall Label Issues", "Inverse Label Issues",
+        "Overall Label Issues", "Inverse Label Noise Score", "Overall Label Quality Score" and a
+        description of each below. Noise scores are between 0 and 1, where 0 implies no label issues
+        in the class. The "Overall Label Quality Score" is also between 0 and 1 where 1 implies
+        perfect quality.
+        Columns:
+        * "Class Index"
+            The index of the class in 0, 1, ..., num_classes - 1.
+        * Overall Label Issues
+            count(given_label = k, true_label != k)
+            Estimated number of label issues in the class (usually the most accurate method).
+        * Inverse Label Issues
+            count(given_label != k, true_label = k)
+            Estimated number of times this class occurs in other classes in the dataset.
+        * Overall Label Noise Score
+            prob(true_label != k | given_label = k)
+            Estimated proportion of label issues in the class.
+        * Inverse Label Noise Score
+            prob(given_label != k | true_label = k)
+            Estimated proportion of times this class label occurs in other classes in the dataset.
+        * Overall Label Quality Score
+            p(given_label = k | true_label = k)
+            How accurate the labels are in each class as a conditional prob p(given=k | true=k).
```

Comment:
Do we really need all 3 of these?  I found it a little overwhelming to see so many scores whose definition I didn't know as a user.

Reply:
At least "Overall Label Quality Score" and "Overall Label Noise Score" convey the same information AFAICT. Would just keep "Overall Label Quality Score" to stick with same convention as rest of the package.

Reply:
"Inverse Label Noise Score" is derivable from "Inverse Label Issues" and the other columns. So want to ask ourself, is precomputing this score for the user worth potentially overwhelming them a little?  I felt a little  overwhelmed by seeing so many scores.

Reply:
Finally anytime the nebulous term "Score" is used, we probably want to stick with convention throughout that lower score -> more serious problem that warrants user's attention.  It is major QOL improvement for user if they don't have to remember which scores are higher = better vs lower = better.

Reply:
Label Quality Score can be very clearly defined as follows:

How accurate the labels tend to be overall for a given class, i.e. the estimated probability that the true label should be `k` averaged across all examples whose given label = `k`.